mypy/server/deps.py

```diff
@@ -151,24 +290,195 @@ def visit_call_expr(self, e: CallExpr) -> None:
+                self.add_operator_method_dependency(right, '__cmp__')
+
+    def process_binary_op(self, op: str, left: Expression, right: Expression) -> None:
+        method = op_methods.get(op)
+        if method:
+            if op != 'in':
+                self.add_operator_method_dependency(left, method)
+            else:
+                self.add_operator_method_dependency(right, method)
+            rev_method = reverse_op_methods.get(method)
+            if rev_method:
+                self.add_operator_method_dependency(right, rev_method)
+
+    def add_operator_method_dependency(self, e: Expression, method: str) -> None:
+        typ = self.type_map.get(e)
+        if typ is not None:
+            self.add_operator_method_dependency_for_type(typ, method)
+
+    def add_operator_method_dependency_for_type(self, typ: Type, method: str) -> None:
+        # Note that we operator methods can't be (non-metaclass) methods of type objects.
+        # TODO: TypedDict
+        # TODO: metaclasses
+        if isinstance(typ, TypeVarType):
+            typ = typ.upper_bound
+        if isinstance(typ, TupleType):
+            typ = typ.fallback
+        if isinstance(typ, Instance):
+            trigger = make_trigger(typ.type.fullname() + '.' + method)
+            self.add_dependency(trigger)
+        elif isinstance(typ, UnionType):
+            for item in typ.items:
+                self.add_operator_method_dependency_for_type(item, method)
+
+    def visit_generator_expr(self, e: GeneratorExpr) -> None:
+        super().visit_generator_expr(e)
+        for seq in e.sequences:
+            self.add_iter_dependency(seq)
+
+    def visit_dictionary_comprehension(self, e: DictionaryComprehension) -> None:
+        super().visit_dictionary_comprehension(e)
+        for seq in e.sequences:
+            self.add_iter_dependency(seq)
+
+    def visit_star_expr(self, e: StarExpr) -> None:
+        super().visit_star_expr(e)
+        self.add_iter_dependency(e.expr)
+
+    def visit_yield_from_expr(self, e: YieldFromExpr) -> None:
+        super().visit_yield_from_expr(e)
+        self.add_iter_dependency(e.expr)
+
     # Helpers
 
     def add_dependency(self, trigger: str, target: Optional[str] = None) -> None:
+        """Add depedency from trigger to a target.
+
+        If the target is not given explicitly, use the current target.
+        """
         if target is None:
             target = self.current()
+        if trigger.startswith(('<builtins.', '<typing.')):
+            # Don't track dependencies to certain builtins to keep the size of
+            # the dependencies manageable. These dependencies should only
+            # change on mypy version updates, which will require a full rebuild
+            # anyway.
+            return
         self.map.setdefault(trigger, set()).add(target)
 
+    def add_type_dependencies(self, typ: Type, target: Optional[str] = None) -> None:
```

Comment:
TODO: Use this in more places where get_type_dependencies() + add_dependency() are called together.